validation.js

```diff
@@ -0,0 +1,296 @@
+const { getMsgIdBFE } = require('./get-msg-id')
+
+function validate(nativeMsg, prevNativeMsg, hmacKey, cb) {
+  let err
+  if ((err = validateSync(nativeMsg, prevNativeMsg, hmacKey))) return cb(err)
+  if ((err = _validateSignature(nativeMsg, hmacKey))) return cb(err)
+  cb()
+}
+
+function validateBatch(nativeMsgs, prevNativeMsg, hmacKey, cb) {
+  let err
+  if ((err = validateBatchSync(nativeMsgs, prevNativeMsg, hmacKey))) {
+    return cb(err)
+  }
+}
+
+function validateBatchSync(nativeMsgs, prevNativeMsg, hmacKey) {
+  let err
+  for (let i = 0; i < nativeMsgs.length; ++i) {
+    const nativeMsg = nativeMsgs[i]
+    if ((err = validateSync(nativeMsg, prevNativeMsg, hmacKey))) return err
+    prevNativeMsg = nativeMsg
+  }
+
+  // FIXME: maybe some random element?
+  const lastNativeMsg = nativeMsgs[nativeMsgs.length - 1]
+  if ((err = _validateSignature(lastNativeMsg, hmacKey))) return err
+}
+
+function validateSync(nativeMsg, prevNativeMsg, hmacKey) {
+  let err
+  if ((err = _validateShape(nativeMsg))) return err
+  if ((err = _validateHmac(hmacKey))) return err
+  if ((err = _validateSize(nativeMsg))) return err
+
+  const [encodedVal, sigBuf, contentBuf] = extract(nativeMsg)
+  const [
+    authorBFE,
+    parentBFE,
+    sequence,
+    timestamp,
+    previousBFE,
+    tag,
+    contentSize,
+    contentHash,
+  ] = extractVal(encodedVal)
+
+  if ((err = _validateTag(tag))) return err
+  if ((err = _validateSequence(sequence))) return err
+  if ((err = _validateContent(contentBuf))) return err
+  if ((err = _validateContentSize(contentSize, contentBuf))) return err
+  if ((err = _validateContentHash(contentHash, contentBuf))) return err
+
+  if ((err = _validateAuthor(authorBFE))) return err
+  if ((err = _validateParent(parentBFE))) return err
+  if ((err = _validateTimestamp(timestamp))) return err
+  if (sequence === 1) {
+    if ((err = _validateFirstPrevious(previousBFE, prevNativeMsg))) return err
+  } else {
+    if ((err = _validatePrevious(encodedVal, prevNativeMsg))) return err
+  }
+}
+
+function _validateShape(nativeMsg) {
+  if (!Buffer.isBuffer(nativeMsg)) {
+    return new Error(`invalid message: expected a buffer`)
+  }
+  const type = bipf.getEncodedType(nativeMsg)
+  if (type !== bipf.types.array) {
+    return new Error(`invalid message: expected first layer to be an array`)
+  }
+}
+
+function _validateHmac(hmacKey) {
+  if (!hmacKey) return
+  if (typeof hmacKey !== 'string' && !Buffer.isBuffer(hmacKey)) {
+    return new Error('invalid hmac key: must be a string or buffer')
+  }
+  const bytes = Buffer.isBuffer(hmacKey)
+    ? hmacKey
+    : Buffer.from(hmacKey, 'base64')
+
+  if (typeof hmacKey === 'string' && bytes.toString('base64') !== hmacKey) {
+    return new Error('invalid hmac')
+  }
+
+  if (bytes.length !== 32) {
+    return new Error('invalid hmac, it should have 32 bytes')
+  }
+}
+
+function _validateSize(nativeMsg) {
+  if (nativeMsg.length > 8192) {
```

Comment:
The spec says 16384. I once asked Dominic about the size and he said, no matter what number you pick they will always ask for more, but I was thinking with something like crut that does a lot of tangle stuff, the 8k can be a bit limiting, so why now bump it just a bit.

Reply:
Fixed